apollo-federation/src/query_plan/fetch_dependency_graph.rs

```diff
@@ -955,23 +978,482 @@ impl FetchDependencyGraph {
+                let Ok(condition_in_supergraph) = self
+                    .supergraph_schema
+                    .get_type(condition.type_name().clone())
+                else {
+                    let condition_name = condition.type_name();
+                    return Err(FederationError::internal(format!(
+                        "Type {condition_name} should exists in the supergraph"
+                    )));
+                };
+                // Note that we're checking the true supergraph, not the API schema, so even @inaccessible types will be found.
+                match condition_in_supergraph {
+                    TypeDefinitionPosition::Interface(interface_type) => Ok(Some(interface_type)),
+                    _ => Ok(None),
+                }
+            } else {
+                Ok(None)
+            }
+        };
+
+        // This condition is specific to the case where we're resolving the _concrete_
+        // `__typename` field of an interface when coming from an interfaceObject type.
+        // i.e. { ... on Product { __typename id }} => { ... on Product { __typename} }
+        // This is usually useless at a glance, but in this case we need to actually
+        // keep this since this is our only path to resolving the concrete `__typename`.
+        let is_interface_type_condition_on_interface_object = |selection: &Selection| {
+            let Some(condition) = try_get_type_condition(selection) else {
+                return Ok::<_, FederationError>(false);
+            };
+            if condition.is_interface_type() {
+                // Lastly, we just need to check that we're coming from a subgraph
+                // that has the type as an interface object in its schema.
+                Ok(self.parents_of(node_index).any(|p| {
+                    let Ok(p_node) = self.node_weight(p) else {
+                        return false;
+                    };
+                    let p_subgraph_name = &p_node.subgraph_name;
+                    let Ok(p_subgraph_schema) = get_subgraph_schema(p_subgraph_name) else {
+                        return false;
+                    };
+                    let Ok(type_in_parent) =
+                        p_subgraph_schema.get_type(condition.type_name().clone())
+                    else {
+                        return false;
+                    };
+                    type_in_parent.is_interface_object_type(p_subgraph_schema.schema())
+                }))
+            } else {
+                Ok(false)
+            }
+        };
+
+        let input_selections: Vec<&Selection> = self_inputs
+            .selection_sets_per_parent_type
+            .values()
+            .flat_map(|s| s.selections.values())
+            .collect();
+        // Checks that every selection is contained in the input selections.
+        node.selection_set
+            .selection_set
+            .iter()
+            .try_fold(true, |acc, selection| {
```

Comment:
Because this `try_fold()` is meant to implement JS `.every()`, you'll want to avoid executing the rest of this function when `acc` is `false`. The easiest thing to do is to just early return `false` here if `acc` is `false`. (Note that this isn't completely short-circuiting since you'll still end up iterating through the rest of the selections in the selection set; one possibility is to extract the code in `try_fold()` into a closure that returns `Result<bool, FederationError>`, and then change that into e.g. `ControlFlow<Result<(), FederationError>>` when calling that closure in the `try_fold()`'s closure, although this is a minor optimization I think.)